### PR TITLE
Solution branch

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,121 @@
...
+        elementsAmount++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > elementsAmount || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bound");
+        }
+        Object[]newArray = new Object[arrayLength];
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:31]: don't create the new Array 
```suggestion
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,121 @@
...
+        System.arraycopy(array, INDEX_NULL, newArray, INDEX_NULL, elementsAmount);
+        System.arraycopy(array, index, newArray, index + 1, elementsAmount - index);
+        newArray[index] = value;
+        array = newArray;
+        elementsAmount++;
+        if (elementsAmount == arrayLength) {
+            resize(arrayLength);
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:39]: move this verification after checking the index
```suggestion
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,121 @@
...
     @Override
     public void addAll(List<T> list) {
-
+        if ((list.size() + elementsAmount) > arrayLength) {
+            resize(elementsAmount + list.size());
+        }
+        for (int i1 = 0; i1 < list.size(); i1++) {
+            array[elementsAmount] = list.get(i1);
+            elementsAmount++;
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:50]:
```suggestion
        for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,121 @@
...
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= elementsAmount || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bound");
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:57]: move this check to private method checkIndex(int index)
```suggestion
        if (index >= elementsAmount || index < 0) {
            throw new ArrayListIndexOutOfBoundsException("Index is out of bound: " + index);
        }
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,121 @@
...
+        return firstElement == null && secondElement == null
+                || firstElement != null && firstElement.equals(secondElement);
+    }
+
+    public void resize(int actualLength) {
+        arrayLength = actualLength + DEFAULT_ARRAY_SIZE / 2;
+        Object[]newArray = new Object[arrayLength];
+        System.arraycopy(array, INDEX_NULL, newArray, INDEX_NULL, elementsAmount);
+        array = newArray;
     }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:111]: Methods that use only in class should be private

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,121 @@
...
     public void set(T value, int index) {
-
+        if (index >= elementsAmount || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bound");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:70]: You can do it simpler
```
    @Override
    public T remove(int index) {
        checkIndex(index);
        T removedElement = array[index];
        System.arraycopy(array, index + 1, array, index, size - 1 - index);
        size--;
        return removedElement;
    }
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,121 @@
...
+        System.arraycopy(array, INDEX_NULL, newArray, INDEX_NULL, elementsAmount);
+        elementsAmount--;
+        System.arraycopy(array, index + 1, newArray, index, elementsAmount - (index));
+        T indexElement = (T) array[index];
+        array = newArray;
+        return indexElement;
     }
 
     @Override
     public T remove(T element) {
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:84]: try to find the index of the `element` and call the method `remove` by index
use one `loop` and one `if`

**Review by Ivan95kos — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,121 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_ARRAY_SIZE = 10;
+    private static final int INDEX_NULL = 0;
+    private Object [] array;
+    private int arrayLength;
+    private int elementsAmount;
```
[the remark this answers is not in the document]
Comment (VadimKadasiev) [reply src/main/java/core/basesyntax/ArrayList.java:10]: Нi, Ivan! 
I haven't found a way to combine these variables. Each of them performs its own function. first is controlling actual array size, but second counts added elements

[the remark this answers is not in the document]
Comment (Rommelua) [reply src/main/java/core/basesyntax/ArrayList.java:10]: arrayLength is duplication
use `array.length` instead

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,109 @@
...
+    private T [] array;
+    private int arrayLength;
+    private int elementsAmount;
+
+    public ArrayList() {
+        arrayLength = DEFAULT_ARRAY_SIZE;
+        array = (T[]) new Object [arrayLength];
+    }
+
+    private void checkIndex(int index) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:17]: Private ones (methods, classes) should be after public ones.

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,109 @@
...
+        if (index > elementsAmount || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bound: " + index);
+        }
+        if (elementsAmount == arrayLength) {
+            resize(arrayLength);
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:23]: `resize `is unexpected to have into `checkIndex `method
remove resize from here

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,109 @@
...
+    }
+
+    private boolean equals(T firstElement, T secondElement) {
+        return firstElement == null && secondElement == null
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:27]:
```suggestion
        return firstElement ==  secondElement 
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,48 +1,109 @@
...
 
     @Override
     public T remove(T element) {
-        return null;
+        int foundElementIndex = -1;
+        for (int i = 0; i < elementsAmount; i++) {
+            if (equals(element, array[i])) {
+                foundElementIndex = i;
+                remove(i);
+                break;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:91]:
```suggestion
                return remove(i);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,48 +1,109 @@
...
+            }
+        }
+        if (foundElementIndex == -1) {
+            throw new NoSuchElementException("Can't find element " + element);
+        }
+        return element;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:97]:
```suggestion
            throw new NoSuchElementException("Can't find element " + element);
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/ArrayList.java (hunk 10 above)
Comment (VadimKadasiev) [reply src/main/java/core/basesyntax/ArrayList.java:23]: This was mentor Ivan's offer "move this verification after checking the index".  Maybe I misunderstood him. 
Should i move it into "add" method.?

**Review by kozhukhovsky — APPROVED** [review]

**Pull request closed without merging**